core/src/main/java/bisq/core/dao/node/BsqNode.java

```diff
@@ -179,19 +189,87 @@ protected void onParseBlockChainComplete() {
         log.info("onParseBlockChainComplete");
         parseBlockchainComplete = true;
         daoStateService.onParseBlockChainComplete();
-
-        // log.error("COMPLETED: sb1={}\nsb2={}", BlockParser.sb1.toString(), BlockParser.sb2.toString());
-        // log.error("equals? " + BlockParser.sb1.toString().equals(BlockParser.sb2.toString()));
-        // Utilities.copyToClipboard(BlockParser.sb1.toString() + "\n\n\n" + BlockParser.sb2.toString());
     }
 
     @SuppressWarnings("WeakerAccess")
     protected void startReOrgFromLastSnapshot() {
         daoStateSnapshotService.applySnapshot(true);
-        startParseBlocks();
     }
 
-    protected boolean isBlockAlreadyAdded(RawBlock rawBlock) {
-        return daoStateService.getBlockAtHeight(rawBlock.getHeight()).isPresent();
+
+    protected Optional<Block> doParseBlock(RawBlock rawBlock) throws RequiredReorgFromSnapshotException {
+        // We check if we have a block with that height. If so we return. We do not use the chainHeight as with genesis
+        // height we have no block but chainHeight is initially set to genesis height (bad design ;-( but a bit tricky
+        // to change now as it used in many areas.)
+        if (daoStateService.getBlockAtHeight(rawBlock.getHeight()).isPresent()) {
+            log.info("We have already a block with the height of the new block. Height of new block={}", rawBlock.getHeight());
+            return Optional.empty();
+        }
+
+        try {
+            Block block = blockParser.parseBlock(rawBlock);
+
+            if (pendingBlocks.contains(rawBlock))
+                pendingBlocks.remove(rawBlock);
+
+            // After parsing we check if we have pending blocks we might have received earlier but which have been
+            // not connecting from the latest height we had. The list is sorted by height
+            if (!pendingBlocks.isEmpty()) {
+                // To avoid ConcurrentModificationException we copy the list. It might be altered in the method call
+                ArrayList<RawBlock> tempPendingBlocks = new ArrayList<>(pendingBlocks);
+                for (RawBlock tempPendingBlock : tempPendingBlocks) {
+                    try {
+                        doParseBlock(tempPendingBlock);
```

Comment:
Yes right I think that is wrong. It would iterate multiple times with lists which have been already applied. E.g. if you have 5 items and first recursive call copies the list then removes the current block from the source list and then copy the list again and start another loop with 4 itmes. Once that recursion is done the parent recursion would still iterate the remaining elements but those have been already processed. So either we add a check there if the block is still in the source list or we dont use a loop but only call with the first element (list is sorted) and so we avoid the remaining iterations from the current loop. 
What do you think?